### ISSUE TITLE
[BUG- AutoMM] Training is much slower in 0.6.x compared with 0.5.2 for image classification

### ISSUE THREAD

**Issue by tonyhoo** [autogluon/autogluon#2756]
- [x] I have checked that this bug exists on the latest stable version of AutoGluon
- [x] and/or I have checked that this bug exists on the latest mainline of AutoGluon via source installation

**Describe the bug**
Training is much slower in 0.6.x compared with 0.5.2 for image classification 

**Expected behavior**
User should expect similar training time when `time_limit` is set across versions using the same HP and dataset
**To Reproduce**
* `pip install autogluon==0.6.2`
* Run following codes with different verions:

```python
if __name__ == "__main__":

    from autogluon.multimodal import MultiModalPredictor
    from autogluon import core
    import time

    import warnings
    warnings.filterwarnings('ignore')
    from autogluon.vision import ImageDataset
    train_dataset, _, test_dataset = ImageDataset.from_folders("https://autogluon.s3.amazonaws.com/datasets/shopee-iet.zip")
    print(train_dataset)

    version = core.version.__version__
    predictor = MultiModalPredictor(label="label", path=f"./automm_imgcls/{version}/{time.time()}")
    start = time.time()
    predictor.fit(
        train_data=train_dataset,
        time_limit=60, # seconds
    ) # you can trust the default config, e.g., we use a `swin_base_patch4_window7_224` model
    end = time.time()
    print(f"version {version} takes {end-start} seconds")
```
**Screenshots**
|       | Time (sec, p3.8xl) |
|-------|------------|
| 0.5.2 | 149        |
| 0.6.1 | 263        |
| 0.6.2|  285        |

**Installed Versions**
Which version of AutoGluon are you are using?  
* 0.6.x
* 0.5.2

```python
# Replace this code with the output of the following:
from autogluon.core.utils import show_versions
show_versions()
```

</details>

**Additional context**

**Comment by sxjscience** [comment]
@zhiqiangdon May be we can also consider to disable greedysoup by default so the total training time can be reduced?

**Comment by tonyhoo** [comment]
@zhiqiangdon @sxjscience  do we know what the accuracy impact will be for greedysoup?

**Comment by tonyhoo** [comment]
@liangfu and @suzhoum  found the removing `env.num_workers` recovered the training throughput back to 0.5.2. Thanks for the work. Until fix is patched and released (tentatively in 0.7). A stop gap solution is to explicitly set `env.num_workers=0` in the hyperparameters. See below example:

```python
if __name__ == "__main__":

    from autogluon.multimodal import MultiModalPredictor
    from autogluon import core
    import time

    import warnings
    warnings.filterwarnings('ignore')
    from autogluon.vision import ImageDataset
    train_dataset, _, test_dataset = ImageDataset.from_folders("https://autogluon.s3.amazonaws.com/datasets/shopee-iet.zip")
    print(train_dataset)

    version = core.version.__version__
    predictor = MultiModalPredictor(label="label", path=f"./automm_imgcls/{version}/{time.time()}")
    hyperparameters = {"env.num_workers": 0}
    start = time.time()
    predictor.fit(
        train_data=train_dataset,
        hyperparameters=hyperparameters,
        time_limit=60, # seconds
    ) # you can trust the default config, e.g., we use a `swin_base_patch4_window7_224` model
    end = time.time()
    print(f"version {version} takes {end-start} seconds")
```

**Comment by sxjscience** [comment]
@tonyhoo I'm aligned in changing `env.num_workers` to be 0 by default.

What's your opinion @zhiqiangdon ?

**Comment by zhiqiangdon** [comment]
Interesting. I remember using `env.num_workers=0` makes image-related training slow. Actually, we have been using `env.num_workers=2` since the creation of AutoMM. Let me double-check the issue.

**Comment by Innixma** [comment]
Is the model being trained the same? If it is a different model then maybe it is ok (for example, if it gets higher accuracy)

**Comment by liangfu** [comment]
It is the same model (swin_base_patch4_window7_224) being trained between v0.5.2 and v0.6.x, but the training throughput is significantly slower on multi-gpu setting.

[This thread](https://discuss.pytorch.org/t/w-cudaipctypes-cpp-22-producer-process-has-been-terminated-before-all-shared-cuda-tensors-released-see-note-sharing-cuda-tensors/124445/8) provides a bit more about the context, and the candidate solution is

> Ok, I dove a bit more in to this, and found the issue. In this particular instance I am using a model library where the author unfortunately had added the batch collate function as a method to the model class. Because of this the model self is also transported to the DataLoader workers.
> When the training scripts exits the DataLoader worker processes are killed too, without properly dealing with the Tensors that are copied to the worker.
> I refactored all the batch collating code out of the model class and thus made it independent of the self of the model instance. This resolved the issue, exiting the script is very smooth now!

**Comment by liangfu** [comment]
Basically, there are three solutions for the overhead in multiprocessing:

1. set num_workers=0, this would leave the dataloader on the main process. This could result in slow preprocessing speed, and make preprocessing the bottleneck. See [num-workers](https://pytorch-lightning.readthedocs.io/en/stable/guides/speed.html#num-workers)
2. set strategy='ddp' instead of 'ddp_spawn'. This would improve the efficiency of multiprocessing, but would introduce the risk of not being able to nest multiprocessing.
3. set persistent_workers=True. This would persist the preprocess workers between iterations, reducing the overhead of creating new subprocesses. See [persistent-workers](https://pytorch-lightning.readthedocs.io/en/stable/guides/speed.html#persistent-workers)

Since we could rely on a significant preprocessing, keeping dataloader on main process (option 1) isn't optimal when we are scaling with DDP. Also, we might need more in-depth discussion on option 2. That would leave the best option 3 for now, which has been draft by @zhiqiangdon in #2776 .

**Comment by zhiqiangdon** [comment]
@liangfu is right about the multiprocessing overhead. It seems that Lightning 1.7.x brought some changes that make ddp_spawn slower. Compared to switching to ddp, setting persistent_workers=True is a safer solution for now.

**Comment by Innixma** [comment]
@zhiqiangdon Is this issue resolved via #2776 being merged?

**Comment by zhiqiangdon** [comment]
Yes, issue resolved.
